File: PWParallaxScrollView/PWParallaxScrollView.m
//
//  PWParallaxScrollView.m
//  PWParallaxScrollView
//
//  Created by wpsteak on 13/6/16.
//  Copyright (c) 2013年 wpsteak. All rights reserved.
//

#import "PWParallaxScrollView.h"

static const NSInteger PWInvalidPosition = -1;

@interface PWParallaxScrollView () <UIScrollViewDelegate>

@property (nonatomic, assign) NSInteger numberOfItems;
@property (nonatomic, assign) NSInteger backgroundViewIndex;
@property (nonatomic, assign) NSInteger userHoldingDownIndex;

@property (nonatomic, strong) UIScrollView *touchScrollView;
@property (nonatomic, strong) UIScrollView *foregroundScrollView;
@property (nonatomic, strong) UIScrollView *backgroundScrollView;

@property (nonatomic, strong) UIView *currentBottomView;

@property (nonatomic, assign) NSInteger currentIndex;

- (void)touchScrollViewTapped:(id)sender;

@end

@implementation PWParallaxScrollView

#pragma mark

- (id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self initControl];
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self initControl];
    }
    return self;
}

- (void)setDataSource:(id<PWParallaxScrollViewDataSource>)dataSource
{
    _dataSource = dataSource;
    [self reloadData];
}

- (void)setForegroundScreenEdgeInsets:(UIEdgeInsets)foregroundScrollViewEdgeInsets
{
    _foregroundScreenEdgeInsets = foregroundScrollViewEdgeInsets;
    [_foregroundScrollView setFrame:UIEdgeInsetsInsetRect(self.bounds, _foregroundScreenEdgeInsets)];
}

- (void)initControl
{
    self.backgroundColor = [UIColor blackColor];
    self.clipsToBounds = YES;
    
    self.touchScrollView = [[UIScrollView alloc] initWithFrame:self.bounds];
    _touchScrollView.delegate = self;
    _touchScrollView.pagingEnabled = YES;
    _touchScrollView.backgroundColor = [UIColor clearColor];
    _touchScrollView.contentOffset = CGPointMake(0, 0);
    _touchScrollView.multipleTouchEnabled = YES;
    
    UITapGestureRecognizer *tapGestureRecognize = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(touchScrollViewTapped:)];
    tapGestureRecognize.numberOfTapsRequired = 1;
    [_touchScrollView addGestureRecognizer:tapGestureRecognize];
    
    self.foregroundScrollView = [[UIScrollView alloc] initWithFrame:self.bounds];
    _foregroundScrollView.scrollEnabled = NO;
    _foregroundScrollView.clipsToBounds = NO;
    _foregroundScrollView.backgroundColor = [UIColor clearColor];
    _foregroundScrollView.contentOffset = CGPointMake(0, 0);
    
    self.backgroundScrollView = [[UIScrollView alloc] initWithFrame:self.bounds];
    _backgroundScrollView.pagingEnabled = YES;
    _backgroundScrollView.backgroundColor = [UIColor clearColor];
    _backgroundScrollView.contentOffset = CGPointMake(0, 0);
    
    [self addSubview:_backgroundScrollView];
    [self addSubview:_foregroundScrollView];
    [self addSubview:_touchScrollView];
}

#pragma mark - public method

- (void)moveToIndex:(NSInteger)index
{
    CGFloat newOffsetX = index * CGRectGetWidth(_touchScrollView.frame);
    [_touchScrollView scrollRectToVisible:CGRectMake(newOffsetX, 0, CGRectGetWidth(_touchScrollView.frame), CGRectGetHeight(_touchScrollView.frame)) animated:YES];
}

- (void)prevItem
{
    if (self.currentIndex > 0) {
        [self moveToIndex:self.currentIndex - 1];
    }
}

- (void)nextItem
{
    if (self.currentIndex < _numberOfItems - 1) {
        [self moveToIndex:self.currentIndex + 1];
    }
}

- (void)reloadData
{
    self.backgroundViewIndex = 0;
    self.userHoldingDownIndex = 0;
    self.numberOfItems = [self.dataSource numberOfItemsInScrollView:self];
    
    UIView *contentView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(self.frame) * _numberOfItems, CGRectGetHeight(self.frame))];
    contentView.backgroundColor = [UIColor clearColor];
    
    [_backgroundScrollView setContentOffset:CGPointMake(0, 0) animated:NO];
    [_backgroundScrollView.subviews makeObjectsPerformSelector:@selector(removeFromSuperview)];
    [_backgroundScrollView addSubview:contentView];
    [_backgroundScrollView setContentSize:contentView.frame.size];
    
    [_foregroundScrollView setContentOffset:CGPointMake(0, 0) animated:NO];
    [_foregroundScrollView.subviews makeObjectsPerformSelector:@selector(removeFromSuperview)];
    [_foregroundScrollView setContentSize:CGSizeMake(CGRectGetWidth(_foregroundScrollView.frame) * _numberOfItems, CGRectGetHeight(_foregroundScrollView.frame))];
    
    [_touchScrollView setContentOffset:CGPointMake(0, 0) animated:NO];
    [_touchScrollView setContentSize:contentView.frame.size];
    
    [self loadBackgroundViewAtIndex:0];
    
    for (NSInteger i = 0; i < _numberOfItems; i++) {
        [self loadForegroundViewAtIndex:i];
    }
}

#pragma mark - private method
- (void)touchScrollViewTapped:(id)sender
{
    if ([self.delegate respondsToSelector:@selector(parallaxScrollView:didRecieveTapAtIndex:)]) {
        [self.delegate parallaxScrollView:self didRecieveTapAtIndex:self.currentIndex];
    }
}

- (UIView *)foregroundViewAtIndex:(NSInteger)index
{
    if (index < 0 || index >= _numberOfItems) {
        return nil;
    }
    
    if (![self.dataSource respondsToSelector:@selector(foregroundViewAtIndex:scrollView:)]) {
        return nil;
    }
    
    UIView *view = [self.dataSource foregroundViewAtIndex:index scrollView:self];
    CGRect newFrame = view.frame;
    newFrame.origin.x += index * CGRectGetWidth(_foregroundScrollView.frame);
    [view setFrame:newFrame];
    [view setTag:index];
    
    return view;
}

- (UIView *)backgroundViewAtIndex:(NSInteger)index
{
    if (index < 0 || index >= _numberOfItems) {
        return nil;
    }
    
    UIView *view = [self.dataSource backgroundViewAtIndex:index scrollView:self];
    [view setFrame:CGRectMake(index * CGRectGetWidth(self.frame), 0, CGRectGetWidth(self.frame), CGRectGetHeight(self.frame))];
    [view setTag:index];
    
    return view;
}

- (void)loadForegroundViewAtIndex:(NSInteger)index
{
    UIView *newParallaxView = [self foregroundViewAtIndex:index];
    
    [_foregroundScrollView addSubview:newParallaxView];
}

- (void)loadBackgroundViewAtIndex:(NSInteger)index
{
    [[_backgroundScrollView subviews] makeObjectsPerformSelector:@selector(removeFromSuperview)];
    
    UIView *newTopView = [self backgroundViewAtIndex:index];
    [_backgroundScrollView addSubview:newTopView];
}

- (void)determineBackgroundView:(float)offsetX
{
    CGFloat newCenterX = 0;
    NSInteger newBackgroundViewIndex = 0;
    NSInteger midPoint = CGRectGetWidth(self.frame) * _userHoldingDownIndex;
    
    if (offsetX < midPoint) {
        //moving from left to right
        
        newCenterX = (CGRectGetWidth(self.frame) * _userHoldingDownIndex - offsetX) / 2;
        newBackgroundViewIndex = _userHoldingDownIndex - 1;
    }
    else if (offsetX > midPoint) {
        //moving from right to left
        
        CGFloat leftSplitWidth = CGRectGetWidth(self.frame) * (_userHoldingDownIndex + 1) - offsetX;
        CGFloat rightSplitWidth = CGRectGetWidth(self.frame) - leftSplitWidth;
        
        newCenterX = rightSplitWidth / 2 + leftSplitWidth;
        newBackgroundViewIndex = _userHoldingDownIndex + 1;
    }
    else {
        newCenterX = CGRectGetWidth(self.frame) / 2 ;
        newBackgroundViewIndex = _backgroundViewIndex;
    }
    
    BOOL backgroundViewIndexChanged = (newBackgroundViewIndex == _backgroundViewIndex) ? NO : YES;
    self.backgroundViewIndex = newBackgroundViewIndex;
    
    if (_userHoldingDownIndex >= 0 && _userHoldingDownIndex <= _numberOfItems) {
        if (backgroundViewIndexChanged) {
            [_currentBottomView removeFromSuperview];
            self.currentBottomView = nil;
            
            UIView *newBottomView = [self backgroundViewAtIndex:_backgroundViewIndex];
            self.currentBottomView = newBottomView;
            [self insertSubview:self.currentBottomView atIndex:0];
        }
    }
    
    CGPoint center = CGPointMake(newCenterX, CGRectGetHeight(self.frame) / 2);
    self.currentBottomView.center = center;
}

- (NSInteger)backgroundViewIndexFromOffset:(CGPoint)offset
{
    NSInteger index = (offset.x / CGRectGetWidth(self.frame));
    
    if (index >= _numberOfItems || index < 0) {
        index = PWInvalidPosition;
    }
    
    return index;
}

#pragma mark UIScrollViewDelegate

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    [_backgroundScrollView setContentOffset:scrollView.contentOffset];
    
    CGFloat factor = _foregroundScrollView.contentSize.width / scrollView.contentSize.width;
    [_foregroundScrollView setContentOffset:CGPointMake(factor * scrollView.contentOffset.x, 0)];
    
    CGFloat offsetX = scrollView.contentOffset.x;
    [self determineBackgroundView:offsetX];
    
    CGRect visibleRect;
    visibleRect.origin = scrollView.contentOffset;
    visibleRect.size = scrollView.bounds.size;
    
    CGRect userPenRect;
    CGFloat width = CGRectGetWidth(scrollView.frame);
    userPenRect.origin = CGPointMake(width * self.userHoldingDownIndex, 0);
    userPenRect.size = scrollView.bounds.size;
    
    if (!CGRectIntersectsRect(visibleRect, userPenRect)) {
        if (CGRectGetMinX(visibleRect) - CGRectGetMinX(userPenRect) > 0) {
            self.userHoldingDownIndex = _userHoldingDownIndex + 1;
        }
        else {
            self.userHoldingDownIndex = _userHoldingDownIndex - 1;
        }
        
        [self loadBackgroundViewAtIndex:_userHoldingDownIndex];
    }
    
    CGFloat newCrrentIndex = round(1.0f * scrollView.contentOffset.x / CGRectGetWidth(self.frame));
    
    if(_currentIndex != newCrrentIndex) {
        self.currentIndex = newCrrentIndex;
        
        if([self.delegate respondsToSelector:@selector(parallaxScrollView:didChangeIndex:)]){
            [self.delegate parallaxScrollView:self didChangeIndex:self.currentIndex];
        }
    }
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
{
    if([self.delegate respondsToSelector:@selector(parallaxScrollView:didEndDeceleratingAtIndex:)]){
        [self.delegate parallaxScrollView:self didEndDeceleratingAtIndex:self.currentIndex];
    }
}

#pragma mark hitTest
- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event
{
    for (UIView *subview in _foregroundScrollView.subviews) {
        CGPoint convertedPoint = [self convertPoint:point toView:subview];
        UIView *result = [subview hitTest:convertedPoint withEvent:event];
        
        if ([result isKindOfClass:[UIButton class]]){
            return result;
        }
    }
    
    return [super hitTest:point withEvent:event];
}

@end

File: PWParallaxScrollView/PWParallaxScrollView.swift
//
//  PWParallaxScrollView.swift
//  PWParallaxScrollView-Swift
//
//  Created by wpsteak on 12/15/15.
//  Copyright © 2015 Pin Shih Wang. All rights reserved.
//

import UIKit

private let PWInvalidPosition :Int = -1

@objc
public protocol PWParallaxScrollViewDataSource : NSObjectProtocol {
    func numberOfItemsInScrollView(scrollView: PWParallaxScrollView) -> Int
    
    optional func backgroundViewAtIndex(index: Int, scrollView: PWParallaxScrollView) -> UIView?
    optional func foregroundViewAtIndex(index: Int, scrollView: PWParallaxScrollView) -> UIView?
}


@objc
public protocol PWParallaxScrollViewDelegate : NSObjectProtocol {
    
    optional func parallaxScrollView(scrollView: PWParallaxScrollView, didChangeIndex index: Int)
    optional func parallaxScrollView(scrollView: PWParallaxScrollView, didEndDeceleratingAtIndex index: Int)
    optional func parallaxScrollView(scrollView: PWParallaxScrollView, didRecieveTapAtIndex index: Int)
}

@objc
public class PWParallaxScrollView: UIView, UIScrollViewDelegate {
    weak public var dataSource: PWParallaxScrollViewDataSource? {
        didSet {
            reloadData()
        }
    }
    weak public var delegate: PWParallaxScrollViewDelegate?
    
    public var foregroundScreenEdgeInsets: UIEdgeInsets = UIEdgeInsetsZero {
        didSet {
            foregroundScrollView.frame = UIEdgeInsetsInsetRect(self.bounds, foregroundScreenEdgeInsets)
        }
    }
    
    private var touchScrollView: UIScrollView = UIScrollView()
    private var foregroundScrollView: UIScrollView = UIScrollView()
    private var backgroundScrollView: UIScrollView = UIScrollView()
    
    private var currentBottomView: UIView?
    
    private var numberOfItems: Int = 0
    private var backgroundViewIndex: Int = 0
    private var userHoldingDownIndex: Int = 0
    
    private var currentIndex: Int = 0

    private func setup() {
        self.backgroundColor = UIColor.blackColor()
        self.clipsToBounds = true;

        touchScrollView.frame = self.bounds
        touchScrollView.pagingEnabled = true
        touchScrollView.backgroundColor = UIColor.clearColor()
        touchScrollView.contentOffset = CGPointMake(0, 0)
        touchScrollView.multipleTouchEnabled = true
        touchScrollView.delegate = self
        
        foregroundScrollView.scrollEnabled = false
        foregroundScrollView.clipsToBounds = false
        foregroundScrollView.backgroundColor = UIColor.clearColor()
        foregroundScrollView.contentOffset = CGPointMake(0, 0);
        
        backgroundScrollView.frame = self.bounds
        backgroundScrollView.pagingEnabled = true
        backgroundScrollView.backgroundColor = UIColor.clearColor()
        backgroundScrollView.contentOffset = CGPointMake(0, 0)
        
        self.addSubview(backgroundScrollView)
        self.addSubview(foregroundScrollView)
        self.addSubview(touchScrollView)
    }
    
    override public init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }

    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    public func reloadData() {
        backgroundViewIndex = 0
        userHoldingDownIndex = 0
        numberOfItems = self.dataSource?.numberOfItemsInScrollView(self) ?? 0
        
        
        let contentView: UIView = UIView(frame: CGRectMake(0, 0, CGRectGetWidth(self.frame) * CGFloat(numberOfItems), CGRectGetHeight(self.frame)))
        contentView.backgroundColor = UIColor.clearColor()
        
        backgroundScrollView.setContentOffset(CGPointMake(0, 0), animated: false)
        backgroundScrollView.subviews.forEach { subview in
            subview.removeFromSuperview()
        }
        backgroundScrollView.addSubview(contentView)
        backgroundScrollView.contentSize = contentView.frame.size
        
        foregroundScrollView.setContentOffset(CGPointMake(0, 0), animated: false)
        foregroundScrollView.subviews.forEach { subview in
            subview.removeFromSuperview()
        }
        foregroundScrollView.contentSize = CGSizeMake(CGRectGetWidth(foregroundScrollView.frame) * CGFloat(numberOfItems), CGRectGetHeight(foregroundScrollView.frame))
        
        touchScrollView.setContentOffset(CGPointMake(0, 0), animated: false)
        touchScrollView.contentSize = contentView.frame.size
        
        loadBackgroundViewAtIndex(0)
        
        for index in 0..<numberOfItems {
            loadForegroundViewAtIndex(index)
        }
    }
    
    public func moveToIndex(index: Int) {
        let newOffsetX: CGFloat = CGFloat(index) * CGRectGetWidth(touchScrollView.frame)
        
        touchScrollView.scrollRectToVisible(CGRectMake(newOffsetX, 0, CGRectGetWidth(touchScrollView.frame), CGRectGetHeight(touchScrollView.frame)), animated: true)
    }
    
    public func prevItem() {
        if (self.currentIndex > 0) {
            moveToIndex(self.currentIndex - 1)
        }
    }

    public func nextItem() {
        if (self.currentIndex < numberOfItems - 1) {
            moveToIndex(self.currentIndex + 1)
        }
    }
    
    private func loadForegroundViewAtIndex(index: Int) {
        if let newParallaxView: UIView = foregroundViewAtIndex(index) {
            foregroundScrollView.addSubview(newParallaxView)
        }
    }

    private func loadBackgroundViewAtIndex(index: Int) {
        backgroundScrollView.subviews.forEach { subview in
            subview.removeFromSuperview()
        }
        
        if let newTopView: UIView = backgroundViewAtIndex(index) {
            backgroundScrollView.addSubview(newTopView)
        }
    }
    
    private func foregroundViewAtIndex(index: Int) -> UIView? {

        if (index < 0 || index >= numberOfItems) {
            return nil
        }
        
        if let view: UIView = self.dataSource?.foregroundViewAtIndex?(index, scrollView: self) {
            var newFrame: CGRect = view.frame
            newFrame.origin.x += CGFloat(index) * CGRectGetWidth(foregroundScrollView.frame)
            view.frame = newFrame
            view.tag = index
            
            return view
        }
        else {
            return nil
        }
    }
    
    private func backgroundViewAtIndex(index: Int) -> UIView? {
        
        if (index < 0 || index >= numberOfItems) {
            return nil
        }
        
        let view: UIView? = self.dataSource?.backgroundViewAtIndex?(index, scrollView: self)
        
        view?.frame = CGRectMake((CGFloat)(index * Int(CGRectGetWidth(self.frame))), 0, CGRectGetWidth(self.frame), CGRectGetHeight(self.frame))
        view?.tag = index
        
        return view
    }
    
    private func determineBackgroundView(offsetX: CGFloat) {
        var newCenterX: CGFloat = 0
        var newBackgroundViewIndex: Int = 0;
        let midPoint: CGFloat = CGRectGetWidth(self.frame) * CGFloat(userHoldingDownIndex);
        
        if (offsetX < midPoint) {
            //moving from left to right
            
            newCenterX = (CGRectGetWidth(self.frame) * CGFloat(userHoldingDownIndex) - offsetX) / 2;
            newBackgroundViewIndex = userHoldingDownIndex - 1;
        }
        else if (offsetX > midPoint) {
            //moving from right to left
            
            let leftSplitWidth: CGFloat = CGRectGetWidth(self.frame) * CGFloat(userHoldingDownIndex + 1) - offsetX;
            let rightSplitWidth: CGFloat = CGRectGetWidth(self.frame) - leftSplitWidth;
            
            newCenterX = rightSplitWidth / 2 + leftSplitWidth;
            newBackgroundViewIndex = userHoldingDownIndex + 1;
        }
        else {
            newCenterX = CGRectGetWidth(self.frame) / 2 ;
            newBackgroundViewIndex = backgroundViewIndex;
        }
        
        let backgroundViewIndexChanged: Bool = (newBackgroundViewIndex == backgroundViewIndex) ? false : true;
        backgroundViewIndex = newBackgroundViewIndex;
        
        if (userHoldingDownIndex >= 0 && userHoldingDownIndex <= numberOfItems) {
            if (backgroundViewIndexChanged) {
                currentBottomView?.removeFromSuperview()
                
                if let newBottomView: UIView = backgroundViewAtIndex(backgroundViewIndex) {
                    currentBottomView = newBottomView;
                    self.insertSubview(newBottomView, atIndex: 0)
                }
            }
        }
        
        let center: CGPoint = CGPointMake(newCenterX, CGRectGetHeight(self.frame) / 2);
        currentBottomView?.center = center;
    }
    
    private func backgroundViewIndexFromOffset(offset: CGPoint) -> Int {
        var index: Int = Int(offset.x / CGRectGetWidth(self.frame))
        
        if (index >= numberOfItems || index < 0) {
            index = PWInvalidPosition;
        }
        
        return index;
    }
    
// MARK: UIScrollViewDelegate
    
    public func scrollViewDidScroll(scrollView: UIScrollView) {
        backgroundScrollView.contentOffset = scrollView.contentOffset
        
        let factor = foregroundScrollView.contentSize.width / scrollView.contentSize.width
        foregroundScrollView.contentOffset = CGPointMake(factor * scrollView.contentOffset.x, 0)
        
        let offsetX = scrollView.contentOffset.x
        determineBackgroundView(offsetX)
        
        let visibleRect = CGRect(origin: scrollView.contentOffset, size: scrollView.bounds.size)
        
        let width: CGFloat = CGRectGetWidth(scrollView.frame);
        let userPenRect = CGRect(origin: CGPointMake(CGFloat(Int(width) * userHoldingDownIndex), 0), size: scrollView.bounds.size)
        
        if (!CGRectIntersectsRect(visibleRect, userPenRect)) {
            if (CGRectGetMinX(visibleRect) - CGRectGetMinX(userPenRect) > 0) {
                userHoldingDownIndex = userHoldingDownIndex + 1
            }
            else {
                userHoldingDownIndex = userHoldingDownIndex - 1
            }
            
            loadBackgroundViewAtIndex(userHoldingDownIndex)
        }
        
        let newCrrentIndex = Int(1.0 * scrollView.contentOffset.x / CGRectGetWidth(self.frame));
        
        if(currentIndex != newCrrentIndex) {
            self.currentIndex = newCrrentIndex;
            
            self.delegate?.parallaxScrollView?(self, didChangeIndex: currentIndex)
        }
    }
    
}
    
